spec/faas/ruby-sam-app/mongodb/app.rb

```diff
@@ -0,0 +1,115 @@
+# frozen_string_literal: true
+
+require 'mongo'
+require 'json'
+
+$open_connections = 0
+$heartbeats_count = 0
+$total_heartbeat_time = 0
+$commands_count = 0
+$total_command_time = 0
+
+
+class CommandMonitor
+  def started(event)
+  end
+
+  def failed(event)
+    $commands_count += 1
+    $total_command_time += event.duration
+  end
+
+  def succeeded(event)
+    $commands_count += 1
+    $total_command_time += event.duration
+  end
+end
+
+class HeartbeatMonitor
+  def started(event)
+  end
+
+  def succeeded(event)
+    $heartbeats_count += 1
+    $total_heartbeat_time += event.duration
+  end
+
+  def failed(event)
+    $heartbeats_count += 1
+    $total_heartbeat_time += event.duration
+  end
+end
+
+class PoolMonitor
+  def published(event)
+    case event
+    when Mongo::Monitoring::Event::Cmap::ConnectionCreated
+      $open_connections += 1
+    when Mongo::Monitoring::Event::Cmap::ConnectionClosed
+      $open_connections -= 1
+    end
+  end
+end
+
+command_monitor = CommandMonitor.new
+heartbeat_monitor = HeartbeatMonitor.new
+pool_monitor = PoolMonitor.new
+
+sdam_proc = proc do |client|
+  client.subscribe(Mongo::Monitoring::COMMAND, command_monitor)
+  client.subscribe(Mongo::Monitoring::SERVER_HEARTBEAT, heartbeat_monitor)
+  client.subscribe(Mongo::Monitoring::CONNECTION_POOL, pool_monitor)
+end
+
+$client = Mongo::Client.new(ENV['MONGODB_URI'], sdam_proc: sdam_proc)
+
+puts 'Connecting'
+$client.use('lambda_test').database.list_collections
+puts 'Connected'
+
+def reset_counters
+  $heartbeats_count = 0
+  $total_heartbeat_time = 0
+  $commands_count = 0
+  $total_command_time = 0
+  $open_connections = 0
+end
+
+def average_heartbeat_time
+  if $heartbeats_count == 0
+    0
+  else
+    $total_heartbeat_time / $heartbeats_count
+  end
+end
+
+def average_command_time
+  if $commands_count == 0
+    0
+  else
+    $total_command_time / $commands_count
+  end
+end
+
+def lambda_handler(event:, context:)
+  db = $client.use('lambda_test')
```

Comment:
Does `$client` need to be a global, here? Is there any reason not to instantiate the client within the handler function?

Reply:
It is [recommended](https://www.mongodb.com/docs/atlas/manage-connections-aws-lambda/) to have a global client instance:
> Don't define a new MongoClient object each time you invoke your function. Doing so causes the driver to create a new database connection with each function call. This can be expensive and can result in your application exceeding database connection limits.

So, we need some kind of the global object here. We can work it around, i.e. using a singleton, but I'd keep it simple and dumb :-)

Reply:
Ah! I didn't know about that constraint. Never mind me, then. 👍 Looks good!